1-Draft/RFCnnnn-Alias-Policy.md

```diff
@@ -0,0 +1,108 @@
+---
+RFC: nnnn
+Author: Ilya Sazonov
+Status: Draft
+Area:
+Comments Due: 6/30/2018
+---
+
+# The document purpose
+
+The purpose of this document is to establish a new policy for the creation and use of aliases to ensure:
+
+* Reliable, predictable behavior of scripts and modules
+* Increase readability and maintainability of scripts and modules
+* UX enhancements in interactive sessions
+
+# Problem description
+
+1. Windows PowerShell has many aliases that conflict with the names of standard Unix utilities.
+
+    These aliases was removed from PowerShell Core.
+    This causes a backward compatibility issue when Windows users do not find the familiar aliases in the interactive session.
+    It also disrupts the operation of custom scripts that use these aliases.
+
+1. Now there are dozens of modules developed by MSFT and thousands of modules of third developers. This causes a name conflict.
+
+    Developers try to create unique cmdlet names.
+    They can use module prefixes to resolve name conflicts.
+
+    On the other hand, developers of the best intentions can complement their modules cmdlet aliases.
+    At the same time, users can create their own aliases.
+
+    With a large number of existing modules and cmdlets and a wide variety of custom aliases, these short names inevitably conflict with each other.
+
+    Moreover, the behavior becomes unpredictable because each subsequent download module or script can change any alias.
+    Either the user can destroy a script behavior by creating a new alias or by deleting an existing one.
+
+    It is even more unpredictable and dangerous to create scripts in the assumption that there are some predefined (standard) aliases because they can be changed by the user or script at any time.
+
+    Modern versions of ScriptAnalyzer already recommend replacing aliases in scripts with their full names.
+
+    >The definition of aliases in modules or scripts and their automatic import becomes an undesirable and dangerous practice.
+
+1. UX in interactive sessions.
+
+    Having many cmdlets and parameters with long names can make it difficult to enter them in interactive sessions.
+    Users who typically use standard settings are typically satisfied with the ```IntellySense``` capabilities.
```

Comment:
## This is an utterly false premise.

1. Users do not have IntelliSense in the console (they have TabExpansion, which is not the same, **at all**).
2. Users are not "typically satisfied" without aliases.

Users use aliases all the time. Particularly in the console, but also in scripts and script modules -- as can be easily demonstrated by examination of source code repositories.

Additionally, users complain when you remove aliases. In fact, **that's why this RFC is being written:** because **each time** the PowerShell team makes changes to aliases, the users feel pain and complain -- this, in and of itself, demonstrates users aren't satisfied with typing out the full names of commands.

Reply:
Just want to echo this.  Alias use is huge in the console, and even in writing custom scripts. When teaching new users, unfamiliar with PowerShell, the using aliases can help simplify adoption.

Reply:
We do not have it today, but I think IntelliSense might be a good solution. PSReadLine 2.0 gets us very close! If it can be made to work as you type instead having to press CTRL + SPACE, then it could be used.

With IntelliSense driven, you could have 3 choices when typing an alias in this order:
- Executable when alias matches an executable
- Fully Resolved Name
- Alias

Include the details on which one it resolves to would ensure the right one is picked.

Reply:
TabExpansion is *far* too slow to be used as IntelliSense...

Reply:
Sometimes so is ISE and VS Code.... It is an idea. Of course it isn't useable now, but with development focus, it could be quite useful. In fact, I use ISE as the "terminal" because of Intellisense.

Reply:
IntelliSense can actually be counter productive in some situations, consider the Azure modules.

Reply:
> Just want to echo this. Alias use is huge in the console, and even in writing custom scripts. When teaching new users, unfamiliar with PowerShell, the using aliases can help simplify adoption.

I just wanted to disagree with this. When helping junior staff learn PowerShell or introducing it to completely new learners, using aliases have *always* been a source of confusion.
% and ? have no logical or intuitive connections to Where-Object or ForEach-Object.
FT/GM/FL might be easier to type, but Get-Member encourages new learners to actually internalise when the command does and is for.

> Users are not "typically satisfied" without aliases.

While users might not be typically "satisfied" without aliases, it is hugely annoying that you never know when an alias from one module will block the importing of another module, unless you first run a command to remove all modules. Something which has easily cost me more time than time I have saved writing GCI instead of Get-C{tab}

 > the dis/enabling of aliases would be better handled in a profile
And so by not having any aliases in PowerShell at all, but maybe include a profile.ps1 file that includes a couple of aliases, you can show people how to create their own if needed.